worlds/dark_souls_3/__init__.py

```diff
@@ -185,364 +223,1350 @@ def create_connection(from_region: str, to_region: str):
+                    and not loc.locked
+                    and loc.item.name in names
+                ]
+
+                # Check the game, not the player, because we know how to sort within regions for DS3
+                offworld = self._shuffle([loc for loc in locations if loc.game != "Dark Souls III"])
+                onworld = sorted((loc for loc in locations if loc.game == "Dark Souls III"),
+                                 key=lambda loc: loc.data.region_value)
+
+                # Give offworld regions the last (best) items within a given sphere
+                for location in onworld + offworld:
+                    new_item = self._pop_item(location, item_order)
+                    location.item = new_item
+                    new_item.location = location
+
+        if self.options.smooth_upgrade_items:
+            base_names = {
+                "Titanite Shard", "Large Titanite Shard", "Titanite Chunk", "Titanite Slab",
+                "Titanite Scale", "Twinkling Titanite", "Farron Coal", "Sage's Coal", "Giant's Coal",
+                "Profaned Coal"
+            }
+            smooth_items([item for item in all_item_order if item.base_name in base_names])
+
+        if self.options.smooth_soul_items:
+            smooth_items([
+                item for item in all_item_order
+                if item.souls and item.classification != ItemClassification.progression
+            ])
+
+        if self.options.smooth_upgraded_weapons:
+            upgraded_weapons = [
+                location.item
+                for location in self.multiworld.get_filled_locations()
+                if location.item.player == self.player
+                and location.item.level and location.item.level > 0
+                and location.item.classification != ItemClassification.progression
+            ]
+            upgraded_weapons.sort(key=lambda item: item.level)
+            smooth_items(upgraded_weapons)
+
+    def _shuffle(self, seq: Sequence) -> List:
+        """Returns a shuffled copy of a sequence."""
+        copy = list(seq)
+        self.random.shuffle(copy)
+        return copy
+
+    def _pop_item(
+        self,
+        location: Location,
+        items: List[Union[DS3ItemData, DarkSouls3Item]]
+    ) -> Union[DS3ItemData, DarkSouls3Item]:
+        """Returns the next item in items that can be assigned to location."""
+        # Non-excluded locations can take any item we throw at them. (More specifically, if they can
+        # take one item in a group, they can take any other).
+        if location.progress_type != LocationProgressType.EXCLUDED: return items.pop(0)
+
+        # Excluded locations require filler items.
+        for i, item in enumerate(items):
+            if item.classification == ItemClassification.filler:
+                return items.pop(i)
+
+        # If we can't find a suitable item, give up and assign an unsuitable one.
+        return items.pop(0)
```

Comment:
This should probably check item rules too right?

Reply:
In fact, was location.can_fill(check_access=False) considered for a lot of these?